### DISCUSSION TITLE
How to add my own filed for user? Like a hased-pin

### DISCUSSION THREAD

**Discussion by weathon** [fastapi-users/fastapi-users#628] [Q&A]
I didn't see it in the doc

**Comment by frankie567** [comment]
Hi there!

It's quite straightforward:

* Add your field(s) definition(s) to the User Pydantic models (https://frankie567.github.io/fastapi-users/configuration/model/#define-your-models).
* If you use SQLAlchemy or Tortoise, you'll have to add the corresponding fields to [`UserTable`](https://frankie567.github.io/fastapi-users/configuration/databases/sqlalchemy/#setup-user-table) or  [`UserDB`](https://frankie567.github.io/fastapi-users/configuration/databases/tortoise/#tweak-userdb-model) respectively.

**Reply by weathon** [reply]
But like if a user input a PIN I need to hash it before I put it in. How do I do this process?

**Reply by frankie567** [reply]
The simplest way is to use a pre-root validator in the Pydantic model:

```py
class User(models.BaseUser):
    pass


class UserCreate(models.BaseUserCreate):
    pin_code: str


class UserUpdate(User, models.BaseUserUpdate):
    pin_code: str


class UserDB(User, models.BaseUserDB):
    hashed_pin_code: str

    @root_validator(pre=True)
    def hash_pin_code(cls, values):
        pin_code = values["pin_code"]
        values["hashed_pin_code"] = hash_function(pin_code)
        return values
```

A "pre" validator allows you to do things before the Pydantic parsing logic is done; so you can provide missing values or change them. Ref: https://pydantic-docs.helpmanual.io/usage/validators/#pre-and-per-item-validators

**Reply by weathon** [reply]
Thank you

**Comment by bert2002** [comment]
Hi, I will hijack this thread, I am trying to expand the User with an own field, but it does not show up.

I updated the modals:
```
class User(models.BaseUser):
  team_id = int

class UserCreate(models.BaseUserCreate):
  pass

class UserUpdate(models.BaseUserUpdate):
  team_id = int

class UserDB(User, models.BaseUserDB):
  team_id = int

class UserTable(Base, SQLAlchemyBaseUserTable):
  team_id = Column(Integer)

```
The table gets created correctly, but when I try to get `team_id` through the `User` model I dont get it. E.g.
```
@app.post("/api/{x}")
def get_model(
  source: ModelSource,
  data: SourceData,
  user: User = Depends(current_active_verified_user)
  ):
  print(user.team_id)
```
and what I get as result is:
```
<class 'int'>
```
When I look at the `user` object I dont see the field as well:
```
id=UUID('782dbd53-94f8-4e8e-9a1a-5349b8cb80e1') email='user@example.com' is_active=True is_superuser=False is_verified=True hashed_password='$2b$12$F6yT/baYbGtlrBnMZMcV1.zQtsdzhSG1SJvhbqeU6wTNS9x49MYlu'
```
I guess I am missing something, but I cant figure it out. Any idea what is missing?

Thanks,
bert

**Reply by frankie567** [reply]
Hi @bert2002,

The type of the field should be specified *via* a type hint, not an assignation, like this:

```py
class User(models.BaseUser):
  team_id: int
```

This is how Pydantic model work. I suggest you go through the FastAPI tutorial to understand this: https://fastapi.tiangolo.com/tutorial/body/#create-your-data-model

**Reply by bert2002** [reply]
omg. Thanks for pointing this out @frankie567. I did not see this. Thanks a lot :)
